plugins/out_s3/s3.c

```diff
@@ -1755,198 +1866,1579 @@ static void s3_upload_queue(struct flb_config *config, void *out_context)
+            flb_sds_destroy(uri);
+            return -1;
+        }
+    }
+
+    s3_client = ctx->s3_client;
+    if (s3_plugin_under_test() == FLB_TRUE) {
+        c = mock_s3_call("TEST_PUT_OBJECT_ERROR", "PutObject");
+    }
+    else {
+        ret = create_headers(ctx, final_body_md5, &headers, &num_headers, FLB_FALSE);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "Failed to create headers");
+            flb_sds_destroy(uri);
+            return -1;
+        }
+
+        c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
+                                              uri, body, body_size,
+                                              headers, num_headers);
+        flb_free(headers);
+    }
+    if (c) {
+        flb_plg_debug(ctx->ins, "PutObject http status=%d", c->resp.status);
+        if (c->resp.status == 200) {
+            /*
+             * URI contains bucket name, so we must advance over it
+             * to print the object key
+             */
+            final_key = uri + strlen(ctx->bucket) + 1;
+            flb_plg_info(ctx->ins, "Successfully uploaded object %s", final_key);
+            flb_sds_destroy(uri);
+            flb_http_client_destroy(c);
+
+            return 0;
+        }
+        flb_aws_print_xml_error(c->resp.payload, c->resp.payload_size,
+                                "PutObject", ctx->ins);
+        if (c->resp.data != NULL) {
+            flb_plg_error(ctx->ins, "Raw PutObject response: %s", c->resp.data);
+        }
+        flb_http_client_destroy(c);
+    }
+
+    flb_plg_error(ctx->ins, "PutObject request failed");
+    flb_sds_destroy(uri);
+
+    return -1;
+}
+
+static int cb_s3_upload_blob(struct flb_config *config, void *data)
+{
+    int ret;
+    char *out_buf = NULL;
+    size_t out_size;
+    uint64_t id;
+    uint64_t file_id;
+    uint64_t part_id;
+    uint64_t part_delivery_attempts;
+    uint64_t file_delivery_attempts;
+    off_t offset_start;
+    off_t offset_end;
+    cfl_sds_t file_remote_id = NULL;
+    cfl_sds_t file_destination = NULL;
+    cfl_sds_t file_path = NULL;
+    cfl_sds_t file_tag = NULL;
+    cfl_sds_t part_ids = NULL;
+    cfl_sds_t source = NULL;
+    struct flb_s3 *ctx = data;
+    struct worker_info *info;
+    struct flb_blob_delivery_notification *notification;
+    struct multipart_upload *m_upload;
+    int part_count;
+    int put_object_required;
+    flb_sds_t pre_signed_url;
+
+    info = FLB_TLS_GET(s3_worker_info);
+
+    if (info->active_upload) {
+        flb_plg_trace(ctx->ins, "[worker: file upload] upload already in progress...");
+
+        return 0;
+    }
+
+    if (ctx->blob_db.db == NULL) {
+        return 0;
+    }
+
+    info->active_upload = FLB_TRUE;
+    pre_signed_url = NULL;
+
+    /*
+     * Check if is there any file which has been fully uploaded and we need to commit it with
+     * the Put Block List operation
+     */
+
+    flb_blob_db_lock(&ctx->blob_db);
+
+    while (1) {
```

Comment:
A large portion of these while(1) loops are duplicating code:
- https://github.com/fluent/fluent-bit/blob/leonardo-master-s3-blob/plugins/out_s3/s3.c#L2582-L2645
- https://github.com/fluent/fluent-bit/blob/leonardo-master-s3-blob/plugins/out_s3/s3.c#L2674-L2736

Can we consider moving to a helper method to avoid the duplication?